walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,127 @@
+        return meets;
+    }
+    else {
+        return optMeets;
+    }
   }
+
+    //  Evaluates the ranges of overlap between two lists of TimeRanges and returns a list of the results
+    private List<TimeRange> findOptionalOverlap(List<TimeRange> meets, List<TimeRange> optFreeTime, long requestDuration){
+        List<TimeRange> optMeets = new ArrayList<>();
+        int meetsNum = meets.size();
+        int optFreeTimeNum = optFreeTime.size();
+
+        for (int i = 0; i < meetsNum; i++) {
+            for (int j = 0; j < optFreeTimeNum; j++) {
+
+                int overlapEnd = Math.min(optFreeTime.get(j).end(),meets.get(i).end());
+                int overlapStart = Math.max(optFreeTime.get(j).start(),meets.get(i).start());
+                
+                // overlapping and the duration of overlap is adequate for a meeting 
+                if (meets.get(i).overlaps(optFreeTime.get(j)) && (requestDuration <= (overlapEnd - overlapStart))){
+                    optMeets.add(TimeRange.fromStartEnd(overlapStart, overlapEnd,false));
+                }      
+            }
+        }
+        return optMeets;
+    }
+
+    // Finds the periods in which attendees are involved in events, sorts them and coalesces overlapping periods. 
+    // Returns the resulting list of TimeRanges
+    private List<TimeRange> findBusyPeriods(Collection<Event> events, Set<String> attendees) {
+        int start = TimeRange.START_OF_DAY; 
+        int end = TimeRange.END_OF_DAY;
+        List<TimeRange> busyTimes = new ArrayList<>();
+        
+        // Adds point of start of the day to aid in coalescing of busy chunks
+        busyTimes.add(TimeRange.fromStartEnd(start, start, false));
+
+        // Determines which events have attendees of interest in them and adds that time range to list
+        for (Event event : events) {
+            if (!Collections.disjoint(event.getAttendees(), attendees)) {
+                busyTimes.add(event.getWhen());
+            }
+        }
+
+        // Adds point of end of the day to aid in coalescing busy chunks 
+        busyTimes.add(TimeRange.fromStartEnd(end+1, end+1,false));
+        coalesceTimes(busyTimes);
+
+        return busyTimes;
+    }
+
+    // Coalesces overlapping busy chunks by setting the start to be the earlier of 2 chunks 
+    // and the end to be the later of the 2 - the separate chunks are removed from the list and 
+    // the coalesced chunk is added in its place
+    private void coalesceTimes(List<TimeRange> busyTimes) {
+        
+        Collections.sort(busyTimes, TimeRange.ORDER_BY_START);
+        int busyTimesLength = busyTimes.size();
+
+        for (int i = 1; i < busyTimesLength;) {
+            TimeRange prior = busyTimes.get(i-1);
+            TimeRange current = busyTimes.get(i);
+            if (prior.overlaps(current)) {
+                TimeRange combined = TimeRange.fromStartEnd(prior.start(), Math.max(prior.end(),current.end()),false);
```

Comment:
Small detail: not sure if the linter points this out but there should be a space after commas.